src/model/interactions/mod.rs

```diff
@@ -0,0 +1,167 @@
+pub mod application_command;
+pub mod message_component;
+pub mod ping;
+
+use application_command::ApplicationCommandInteraction;
+use bitflags::__impl_bitflags;
+use message_component::MessageComponentInteraction;
+use ping::PingInteraction;
+use serde::de::{Deserialize, Deserializer, Error as DeError};
+use serde::ser::{Serialize, Serializer};
+use serde_json::Value;
+
+use super::prelude::*;
+use crate::internal::prelude::*;
+
+#[derive(Clone, Debug)]
+pub enum Interaction {
+    Ping(PingInteraction),
+    ApplicationCommand(ApplicationCommandInteraction),
+    MessageComponent(MessageComponentInteraction),
+}
+
+impl Interaction {
+    /// Gets the interaction Id.
+    pub fn id(&self) -> InteractionId {
+        match self {
+            Interaction::Ping(i) => i.id,
+            Interaction::ApplicationCommand(i) => i.id,
+            Interaction::MessageComponent(i) => i.id,
+        }
+    }
+
+    /// Gets the interaction type
+    pub fn kind(&self) -> InteractionType {
+        match self {
+            Interaction::Ping(_) => InteractionType::Ping,
+            Interaction::ApplicationCommand(_) => InteractionType::ApplicationCommand,
+            Interaction::MessageComponent(_) => InteractionType::MessageComponent,
+        }
+    }
+
+    /// Gets the interaction application Id
+    pub fn application_id(&self) -> ApplicationId {
+        match self {
+            Interaction::Ping(i) => i.application_id,
+            Interaction::ApplicationCommand(i) => i.application_id,
+            Interaction::MessageComponent(i) => i.application_id,
+        }
+    }
+
+    /// Gets the interaction token.
+    pub fn token(&self) -> &String {
```

Comment:
Is it useful to return a reference to a `String`? The only scenario this would be convenient is if the user is interested in the capacity of the allocated buffer, but I doubt there is someone that needs to know that. A simpler alternative would be to return `&str`:
```suggestion
    pub fn token(&self) -> &str {
```